Custom user data profiles

orangejenny opened the pull request:
##### SUMMARY
https://trello.com/c/d5QAXui5/85-custom-user-data-profile

Review by commit. Notes on implementation:
- The profile id is stored as a value in `user_data` tied to a new system key.
- Most usage of the `CouchUser.user_data` attribute is replaced with a `metadata` property. For `CouchUser` this just returns `user_data`. For `CommCareUser` it checks for the profile key and, if found, adds in data from the profile. Updates to metadata that affect keys that are already set by the profile will fail.
- ES still uses the `user_data_es` property. There's just a pillow update to look up the profile and add in any values. ~~This does mean it's possible for a couch doc and ES doc to be out of sync: changes to the *profile* will be immediately reflected in the couch document but will not be reflected in ES until the next time the ES doc is updated.~~ Fixed sync issue in 6d8e156060722ba1de01fe59f7d5348f6958999f
- The API also still references `user_data`. The API can be used to update the profile via the system key.

Still TODO:

- [x] Fix ES sync issue, where updating a profile does not update ES user docs
- [x] Fix tests
- [x] Finish [docs](https://confluence.dimagi.com/display/ccinternal/User+field+profiles)

##### FEATURE FLAG
New feature flag: "User field profiles." The flag controls the UI availability, but it has no effect on the user data itself, so if the flag is turned off, projects should continue to function.

##### RISK ASSESSMENT / QA PLAN
[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-1622)

##### PRODUCT DESCRIPTION
This adds "profiles" to custom user fields, which allow saving a set of fields and values and then assigning that set to users. Profiles can be assigned to mobile users when creating or editing in the UI, or via the bulk download/upload.

Profiles are managed in a new tab on the page to edit user fields:
![Screen Shot 2020-07-24 at 3 12 21 PM](https://user-images.githubusercontent.com/1486591/88427423-e05fa180-cdc0-11ea-8a33-ffee5ecbe661.png)

![Screen Shot 2020-07-24 at 3 12 44 PM](https://user-images.githubusercontent.com/1486591/88427435-e3f32880-cdc0-11ea-9b25-4eb5426ec8dc.png)

Assigning users to profiles on editing:
![Screen Shot 2020-07-24 at 3 31 35 PM](https://user-images.githubusercontent.com/1486591/88428422-d474df00-cdc2-11ea-9964-5994e514958b.png)

orangejenny committed:
Added CustomDataFieldsProfile

orangejenny committed:
Added tabbed UI (w=1)

orangejenny committed:
Added UI for creating and editing profiles

orangejenny committed:
Added UI support for adding and saving profiles

orangejenny committed:
Added feature flag

orangejenny committed:
Added linked domain support

orangejenny committed:
Added support for removing profiles from UI

orangejenny committed:
Added validation that profiles only use fields that exist in definitions

- Profiles can only use fields that exist in definitions
- Profile values must pass field validation for choices and regex

orangejenny committed:
Updated profile saving to handle edits and deletions

orangejenny committed:
Added PROFILE_SLUG and support editing in bulk download/upload

orangejenny committed:
Added support for profile to edit mobile worker page

orangejenny committed:
Added support for profile to create mobile worker

orangejenny committed:
Added incompatibility with add_users_from_location flag

This is a separate user creation workflow that's on its way out, not worth supporting.

orangejenny committed:
Updated js to get PROFILE_SLUG from initial page data

orangejenny committed:
Added logic to populate original values

orangejenny committed:
Added CouchUser.metadata property

orangejenny committed:
Added update_metadata and pop_metadata functions

orangejenny committed:
Renamed user_data to metadata in a few places

These aren't strictly necessary, they're just to reinforce the new name

orangejenny committed:
Updated user_data references to use metadata

orangejenny committed:
Added explicit metadata argument to CouchUser.create

orangejenny committed:
Updated transform_user_for_elasticsearch to include profile properties

orangejenny committed:
Added has_users_assigned and prevent deletion

orangejenny committed:
Improved API support

API doesn't support a separate profile field like the UI bulk management,
but it accepts PROFILE_SLUG and will error if an update contains user data
that conflicts with a profile.

orangejenny commented:
This is a big PR that's prone to merge conflicts. In an attempt to keep the commit history clean, until review starts, I'm going to handle conflicts and qabugs by rebasing and force pushing.

calellowitz commented on corehq/apps/custom_data_fields/models.py:
```diff
@@ -142,3 +142,9 @@ def get_model_and_uncategorized(self, data_dict):
+
+class CustomDataFieldsProfile(models.Model):
+    name = models.CharField(max_length=126)
+    fields = JSONField(default=list, null=True)
```
should the default here be a dict?

orangejenny replied:
Yes. 275ea2c199e

orangejenny replied:
Yes. 275ea2c199e

calellowitz commented on corehq/apps/user_importer/importer.py:
```diff
@@ -423,12 +423,14 @@ def _get_domain_info(domain):
-                    for key in domain_info.profiles_by_name[profile].fields.keys():
-                        user.user_data.pop(key, None)   # TODO: pop fields when editing user, too
+                    data[PROFILE_SLUG] = domain_info.profiles_by_name[profile].id
+                if data:
```
Does the metadata setter enforce this now?

orangejenny replied:
The TODO? Yes.

calellowitz reviewed (commented):
Lots of small questions. I am not totally sure I wrapped my head around all the mechanics of this but generally looks good

orangejenny committed:
Switched order of form and user data when editing user (QA-1627)

orangejenny committed:
Fixed translation

orangejenny committed:
Removed lowercasing when checking slug validity

This was copied from the function above but doesn't apply here.

orangejenny commented on corehq/apps/custom_data_fields/edit_model.py:
```diff
@@ -118,6 +118,26 @@ def clean_regex(self):
+    """
+    name = forms.CharField(
+        required=True,
+        error_messages={'required': ugettext_lazy('All fields are required')}
```
Updated in a later commit.

orangejenny commented on corehq/apps/users/views/mobile/custom_data_fields.py:
```diff
@@ -14,7 +14,7 @@ class UserFieldsView(CustomDataModelMixin, BaseUserSettingsView):
     entity_string = _("User")
     show_purge_existing = True
-    show_profiles = True
+    _show_profiles = True
```
Is there a convention for protected?

orangejenny commented on corehq/apps/custom_data_fields/edit_model.py:
```diff
@@ -51,6 +53,42 @@ def verify_no_reserved_words(self, data_fields):
+                 for field in data_fields if 'slug' in field}
+        for profile in profiles:
+            for field in json.loads(profile.get('fields', "{}")).keys():
+                if field not in slugs:
```
I shouldn't have lowercased the slugs, that came up as a bug. Fixed in 5fcc75231d1cbf2ebc77d6bc1aa72bf745aed747

orangejenny commented on corehq/apps/user_importer/importer.py:
```diff
@@ -411,6 +425,10 @@ def _get_domain_info(domain):
                     user.set_full_name(str(name))
                 if data:
                     user.user_data.update(data)
+                if profile:
```
It assigns the profile. The profile id is stored in `metadata` (formerly `user_data`) but unlike other data fields that get a `data: slugname` column in the upload & download, the profile gets its own special column.

orangejenny commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -934,6 +935,14 @@ def dispatch(self, request, *args, **kwargs):
         return super(CreateCommCareUserModal, self).dispatch(request, *args, **kwargs)
 
     def render_form(self, status):
+        if CUSTOM_DATA_FIELDS_PROFILES.enabled(self.domain):
```
It's a button on the edit location page that's shown if you have the add_users_from_location flag turned on.

orangejenny committed:
Switched CustomDataFieldsProfile.fields default from list to dict

snopoke committed:
erge branch 'master' into jls/CustomDataFieldsProfile

snopoke commented:
See also: https://github.com/dimagi/commcare-icds/pull/3

snopoke committed:
Merge branch 'master' into jls/CustomDataFieldsProfile

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into jls/CustomDataFieldsProfile

orangejenny committed:
Merge branch 'master' into jls/CustomDataFieldsProfile

orangejenny committed:
Fixed inability to create user without profile

orangejenny committed:
Fixed conflict check

orangejenny committed:
Added clearing to profile dropdown (QA-1658)

orangejenny committed:
Fixed select2 dropdown not updating on profile change

orangejenny committed:
Fixed header name for bulk user upload

orangejenny committed:
Added user import validator to verify profile names

orangejenny committed:
Updated logic around error display

Always display an overall failure or success message in addition to form-level errors.

orangejenny committed:
Updated save button to be enabled if save failed

orangejenny committed:
Fixed profiles being incompatible with required fields

orangejenny committed:
Fixed select2 behavior

- Updated js dependencies so select2 binding is available
- Made sure profile dropdown is always a select2

orangejenny commented on corehq/apps/custom_data_fields/edit_model.py:
```diff
@@ -118,6 +118,26 @@ def clean_regex(self):
+        required=True,
+        error_messages={'required': ugettext_lazy('All fields are required')}
+    )
+    fields = forms.CharField(
```
It's a hidden input that's part of the modal key/value widget.

orangejenny commented on corehq/apps/custom_data_fields/edit_model.py:
```diff
@@ -118,6 +118,26 @@ def clean_regex(self):
+    def clean_fields(self):
+        fields = self.data.get('fields')
+        if fields:
+            return json.dumps(fields)
```
It isn't really a string. This form is created by the outer `CustomDataFieldsForm` that controls the entire page (field definitions + profiles), in `clean_profiles`, which passes in a dict. It's a weird setup. Would it be clearer to re-encode the fields as.a string when the form is created? Django 3.1 adds a JSONField I'd use if it were available.

orangejenny committed:
Added refresh_es_for_profile_users

orangejenny committed:
Fixed casing in messages

orangejenny committed:
Updated test_inactive_user_query to be like test_active_user_query

orangejenny committed:
Fixed TestCommCareUserResource.test_update to check metadata, not user_data

orangejenny committed:
Switched copy in metadata getter to deepcopy

This prevents metadata being inadvertently overwritten, fixes test failure
TestDemoUser.test_demo_user_data_in_restore that was caused by
turn_on_demo_mode modifying metadata.

orangejenny committed:
WIP: See if travis at 3.7 fixes test failures

orangejenny committed:
Revert "WIP: See if travis at 3.7 fixes test failures"

This reverts commit 7348e80158cdbb255d8ae51810a9d8f8b3e04593.

orangejenny committed:
Updated copy logic in an attempt to fix tests

orangejenny committed:
Added inactive useres when checking if profile has users assigned

orangejenny committed:
Removed fake placeholder choice, leaving placeholder data attribute

orangejenny committed:
Added test that redundant data isn't added to user_data

orangejenny committed:
Renamed user import column

orangejenny committed:
Fixed placeholder for choice fields

orangejenny committed:
Added check that profiles can't have duplicate names

orangejenny committed:
Added profile id to error message for missing profile

orangejenny committed:
Added logic to remember tab on form submit

orangejenny committed:
Fixed new test

proteusvacuum reviewed (approved):
Looks great! I looked mostly at the python code.

orangejenny commented on corehq/apps/api/resources/v0_5.py:
```diff
@@ -244,6 +244,11 @@ def _update(self, bundle):
                     should_save = True
+                elif key == 'user_data':
+                    try:
+                        bundle.obj.metadata = value
```
This is a good point, I've been mulling it over since Friday. I'm inclined to leave it as-is, to follow precedent and because I don't have a good alternative. Projects that use profiles and the API will be subject to errors if they update user_data in a way that conflicts with the profile, so it doesn't seem right to not let them access the profile.

There are other system keys in user data, including the user's location, that are similarly undocumented and have the same risk, which makes me think people either don't run into it or have learned to pull the user_data and then add/remove keys, rather than wholesale replacing it. I think I'll add that as a best practice to the API docs.

proteusvacuum replied:
I think this is sound rationale. I saw your docs update - I think it might be worth mentioning in there the specific keys that can be dangerous to overwrite, or give a pointer to how to "pull" the latest user data.

orangejenny replied:
👍  updated

orangejenny commented on corehq/apps/linked_domain/tests/test_update_custom_data_fields.py:
```diff
@@ -0,0 +1,90 @@
+        self.assertEqual(profiles_by_name[self.fish_profile.name].fields, self.fish_profile.fields)
+
+        # Add, update, and remove a profile
+        lamprey_profile = CustomDataFieldsProfile(
```
😱 

I like their many rows of teeth.

orangejenny committed:
Switched assertEquals to assertEqual

https://github.com/python/cpython/blob/master/Doc/library/unittest.rst#deprecated-aliases

orangejenny commented on corehq/apps/reports/tests/test_esaccessors.py:
```diff
@@ -841,24 +866,34 @@ def _send_user_to_es(self, _id=None, is_active=True):
 
         self.assertEqual(len(results), 1)
+        metadata = results[0].pop('user_data_es')
+        self.assertEquals({
```
Oops, I often mistype that. I thought they were equivalent, but apparently [assertEquals is deprecated](https://github.com/python/cpython/blob/5e3826785dcc64f8e1a8a7bde11b88fbb40943be/Doc/library/unittest.rst#deprecated-aliases). Will fix.

orangejenny committed:
Updated metadata logic to be sure later changes don't affect the original user_data
